File: lorekeeper/external/classifiers.py
"""Classifiers for external events."""

from typing import Iterable, List

from .schemas import ExternalEvent

KEYWORDS = ("ship", "launch", "release", "milestone", "highlight")


def detect_milestone(event: ExternalEvent) -> str | None:
    if event.milestone:
        return event.milestone

    if event.type in {"milestone", "story_highlight", "release"}:
        return event.type

    text = (event.text or "").lower()
    for keyword in KEYWORDS:
        if keyword in text:
            return keyword
    return None


def classify(events: Iterable[ExternalEvent]) -> List[ExternalEvent]:
    milestones: List[ExternalEvent] = []
    for event in events:
        marker = detect_milestone(event)
        if marker:
            event.milestone = marker
            milestones.append(event)
    return milestones

File: apps/server/src/external/classifiers.ts
import { ExternalEvent } from './types';

const milestoneKeywords = ['shipped', 'launch', 'released', 'milestone', 'highlight'];

export function detectMilestone(event: ExternalEvent): string | null {
  if (typeof event.milestone === 'string' && event.milestone.length > 0) {
    return event.milestone;
  }

  if (event.type === 'milestone' || event.type === 'story_highlight' || event.type === 'release') {
    return event.type;
  }

  const text = event.text?.toLowerCase() ?? '';
  const keyword = milestoneKeywords.find((keyword) => text.includes(keyword));
  return keyword ?? null;
}

export function classifyMilestones(events: ExternalEvent[]): ExternalEvent[] {
  return events
    .map((event) => ({ ...event, milestone: detectMilestone(event) }))
    .filter((event) => event.milestone !== null);
}
